pkg/services/grpcserver/datasource.go

```diff
@@ -0,0 +1,131 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/genproto/pluginv2"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/expr"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/adapters"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/tsdb/grafanads"
+	"google.golang.org/grpc"
+)
+
+type DatasourceService struct {
+	cfg    *setting.Cfg
+	server *DatasourceServer
+}
+
+func ProvideDatasourceService(cfg *setting.Cfg, grpcServerProvider Provider, pluginsClient plugins.Client, dataSourceCache datasources.CacheService, dataSourceService datasources.DataSourceService) (*DatasourceService, error) {
+	return &DatasourceService{
+		cfg: cfg,
+		server: &DatasourceServer{
+			pluginClient:      pluginsClient,
+			dataSourceCache:   dataSourceCache,
+			dataSourceService: dataSourceService,
+		},
+	}, nil
+}
+
+func (s *DatasourceService) Run(ctx context.Context) error {
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func (s *DatasourceService) IsDisabled() bool {
+	if s.cfg == nil {
+		return true
+	}
+	return !s.cfg.IsFeatureToggleEnabled(featuremgmt.FlagGrpcServer)
+}
+
+type DatasourceServer struct {
+	pluginClient      plugins.Client
+	dataSourceCache   datasources.CacheService
+	dataSourceService datasources.DataSourceService
+}
+
+// QueryData handles multiple queries and returns multiple responses.
+func (s *DatasourceServer) QueryData(ctx context.Context, in *pluginv2.QueryDataRequest, opts ...grpc.CallOption) (*pluginv2.QueryDataResponse, error) {
+	var ds *models.DataSource
+	req := backend.FromProto().QueryDataRequest(in)
+	user := adapters.SignedInUserFromBackendUser(req.PluginContext.User)
```

Comment:
not sure if this is a bad idea, but i was thinking the `User` and `OrgId` would come from the parent query request that was sent to the app plugin backend.